File: 选择题/适配器模式.cpp
#include <iostream>
using namespace std;
 
// Targets
class Target
{
public:
	virtual void Request()    // Methods  
	{
		cout << "Target::Request" << endl;
	}
};
 
// Adaptee
class Adaptee
{
public:
	void SpecificRequest()   // Methods 
	{
		cout << "Adaptee::SpecificRequest" << endl;
	}
};
 
// Adapter
class Adapter : public Target, Adaptee
{
public:
	void Request()        // Implements ITarget interface
	{
		// Possibly do some data manipulation  
		// and then call SpecificRequest   
		this->Request();
		this->SpecificRequest();
	}
};
 
// Client
int main(int argc, char *argv[])
{
	Target *targetObj = new Adapter();
	targetObj->Request();
 
	delete targetObj;
	targetObj = NULL;
 
	return 0;
}

File: 2-书本练习/Design Pattern/2-结构型模式/1-适配器模式/适配器模式.py
class Target:
    def request(self):
        pass

class Adaptee:
    def specific_request(self):
        print("Adaptee specific request")

class Adapter(Target):
    def __init__(self, adaptee: Adaptee):
        self.adaptee = adaptee

    def request(self):
        self.adaptee.specific_request()

if __name__ == '__main__':
    adaptee = Adaptee()
    adapter = Adapter(adaptee)

    adapter.request()
